### PR TITLE
✨ Implement new amp-access-poool extension

### PR DESCRIPTION
**Pull request by dackmin**
Implements a new `amp-access-poool` extension (see #18699)

Please feel free to comment any line of that PR !

### CODE REVIEW

**Comment by dackmin** [comment]
@googlebot

**Comment by NicolasAuger** [comment]
@googlebot Yes, i'm ok with that 😄

**Comment by Gregable** [comment]
For validator *only*. Looks good, but would be nice to have a test case. Obviously it's trivial, but would catch typos and such.

**Comment by dvoytenko** [comment]
@dackmin Thanks! Reviewing...

File: extensions/amp-access-poool/0.1/poool-impl.js [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+  }
+
+  /**
+   * Produces the Poool SDK object for the passed in callback.
+   *
+   * @param {!Window} global
+   * @param {function(!Object)} cb
+   */
+  getPooolSDK_(global, cb) {
+    loadScript(global, 'https://assets.poool.fr/poool.min.js', function() {
```
Comment (jpettitt) [inline extensions/amp-access-poool/0.1/poool-impl.js]: This is a problem. AMP does not allow 3p JS to be loaded on the main page.

Comment (dackmin) [reply extensions/amp-access-poool/0.1/poool-impl.js]: yeah we noticed that on travis, we will try a different approach, although our paywall is already executed inside an iframe.

Comment (dvoytenko) [reply extensions/amp-access-poool/0.1/poool-impl.js]: @dackmin To quickly add: there are couple of libraries that mostly operate via a 3p iframe. Please see `amp-access-*` and `amp-subscriptions-*` extensions.

**Comment by jpettitt** [comment]
Looks like you still have some test fails.

**Comment by jpettitt** [comment]
@dackmin you need to sign the cla using the email address that's in your commits otherwise the bot is going to keep complaining every time there is a new commit.

**Comment by dackmin** [comment]
@jpettitt pretty sure I already signed it with all the emails I usually commit with, will check with @NicolasAuger if he signed with the right email though.

It also seems the bot is seeing the CLAs as signed but needs authors consent or manual approbation.

**Comment by jpettitt** [comment]
@dackmin - I looked up your github id and it's not showing as signed in our DB.  @NicolasAuger  shows as signed.

**Comment by jpettitt** [comment]
The percy error you are seeing should be fixed by merging the latest master, it's unrelated to your code.

File: extensions/amp-access-poool/0.1/amp-access-poool.css [hunk 2]
```diff
@@ -0,0 +1,15 @@
+/**
```
Comment (jpettitt) [inline extensions/amp-access-poool/0.1/amp-access-poool.css]: Do you need this file - there is no code in it.

File: extensions/amp-access-poool/0.1/poool-impl.js [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+    /** @private {string} */
+    this.itemID_ = this.pooolConfig_['itemID'] || '';
+
+    /** @const {!Element} */
+    this.iframe_ = document.createElement('iframe');
+
+    this.initializeIframe_();
+
+    installStylesForDoc(this.ampdoc, CSS, () => {}, false, TAG);
```
Comment (jpettitt) [inline extensions/amp-access-poool/0.1/poool-impl.js]: You don't seem to have any styles to install, remove if not needed

**Comment by jpettitt** [comment]
Manual override on CLA (poool has a corp CLA on file)

**Comment by dackmin** [comment]
That bot is definitely not my friend.

**Comment by jpettitt** [comment]
@dvoytenko needs to review this to meet the new approvals process, I've pinged him, however it's thanksgiving week here so it may not be until next week.

**Comment by dackmin** [comment]
> The bot will go away if you check that every email mention in a commit has signed the CLA.

Tried to sign it again, we'll see.

> @dvoytenko needs to review this to meet the new approvals process, I've pinged him, however it's thanksgiving week here so it may not be until next week.

Roger that, thanks !

**Comment by jpettitt** [comment]
OK I see you in the signed CLA's now - the bot should be happy going forward.

File: extensions/amp-access-poool/0.1/poool-impl.js [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+  /**
+   * @private
+   */
+  renderPoool_() {
+    const pooolContainer = document.getElementById('poool');
+    const urlPromise = this.accessSource_.buildUrl(this.iframeUrl_
+        + '&bi=' + this.pooolConfig_['bundleID']
+        + '&iid=' + this.pooolConfig_['itemID']
+        + '&ce=' + this.pooolConfig_['cookiesEnabled']
+        + '&d=' + this.pooolConfig_['debug']
```
Comment (dvoytenko) [inline extensions/amp-access-poool/0.1/poool-impl.js]: We usually advise to use `getMode().development` for `debug` kind of things. It turns on a bunch of other debug features as well. Either way is ok, but thought you may want to consider it.

File: extensions/amp-access-poool/0.1/poool-impl.js [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+    user().assert(this.bundleID_, 'BundleID is incorrect or not provided.');
+    user().assert(this.itemID_, 'ItemID is not provided.');
+  }
+
+  /**
+   * @return {!Promise<Object>}
+   * @private
+   */
+  getPooolAccess_() {
+    const url = this.accessUrl_ + '&iid=' + this.itemID_;
```
Comment (dvoytenko) [inline extensions/amp-access-poool/0.1/poool-impl.js]: Please either use `addParamToUrl` or `encodeURIComponent()` for the `this.itemID_` variable (even if it's always expected to be URL-safe).

File: extensions/amp-access-poool/0.1/poool-impl.js [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+    });
+  }
+
+  /**
+   * @private
+   */
+  renderPoool_() {
+    const pooolContainer = document.getElementById('poool');
+    const urlPromise = this.accessSource_.buildUrl(this.iframeUrl_
+        + '&bi=' + this.pooolConfig_['bundleID']
```
Comment (dvoytenko) [inline extensions/amp-access-poool/0.1/poool-impl.js]: Ditto, please use `addParamToUrl` or `encodeURIComponent` for all of the params below.

**Comment by dvoytenko** [comment]
LGTM on my side. Thanks!

**Comment by jpettitt** [comment]
@dackmin @NicolasAuger your PR is now merged.  Congrats on your new extension.

Due to the holiday code freeze it won't make it to production until some time in January.  Look for the "In Canary" and "In Production" labels on this PR to track status.  Once it hits canary you will be able to test it in the real world using the dev channel setting at https://cdn.ampproject.org/experiments.html

File: extensions/amp-access-poool/validator-amp-access-poool.protoascii [hunk 7]
```diff
@@ -0,0 +1,28 @@
+#
+# Copyright 2017 The AMP HTML Authors. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (jpettitt) [reply extensions/amp-access-poool/validator-amp-access-poool.protoascii]: #19928

**Comment by dackmin** [comment]
@dvoytenko @jpettitt any idea when this PR will be released in canary or production?

**Comment by jpettitt** [comment]
It's in production, has been since January.

https://cdn.ampproject.org/v0/amp-access-poool-0.1.js

**Pull request merged**